src/sagas/nanoContract.js

```diff
@@ -0,0 +1,116 @@
+import {
+  ncApi
+} from '@hathor/wallet-lib';
+import {
+  takeEvery,
+  select,
+  all,
+  put,
+  call,
+} from 'redux-saga/effects';
+import { t } from 'ttag';
+import { cloneDeep } from 'lodash';
+import { NanoRequest404Error } from '@hathor/wallet-lib/lib/errors';
+import { STORE } from '../store';
+import {
+  nanoContractRegisterFailure,
+  nanoContractRegisterSuccess,
+  onExceptionCaptured,
+  types,
+} from '../actions';
+
+export const failureMessage = {
+  alreadyRegistered: t`Nano Contract already registered.`,
+  walletNotReadyError: t`Wallet is not ready yet to register a Nano Contract.`,
+  addressNotMine: t`The informed address do not belongs to the wallet.`,
+  nanoContractStateNotFound: t`Nano Contract not found.`,
+  nanoContractStateFailure: t`Error while trying to get Nano Contract state.`,
+};
+
+/**
+ * Calls Nano Contract API to retrieve Nano Contract state.
+ * @param {string} ncId Nano Contract ID
+ * @returns {{
+ *   ncState?: Object,
+ *   error?: Error,
+ * }} Returns either an object containing ncState or an error.
+ */
+export async function getNanoContractState(ncId) {
+  try {
+    const state = await ncApi.getNanoContractState(ncId);
+    return { ncState: { ...state } };
+  } catch (err) {
+    return { error: err };
+  }
+}
+
+/**
+ * Process Nano Contract registration request.
+ * @param {{
+ *   payload: {
+ *     address: string,
+ *     ncId: string,
+ *   }
+ * }} action with request payload.
+ */
+export function* registerNanoContract({ payload }) {
+  const { address, ncId } = payload;
+  const storage = STORE.getStorage();
+
+  // validate address belongs to the wallet
+  const wallet = yield select((state) => state.wallet);
+  if (!wallet.isReady()) {
+    yield put(nanoContractRegisterFailure(failureMessage.walletNotReadyError));
+    // This will show user an error modal with the option to send the error to sentry.
+    yield put(onExceptionCaptured(new Error(failureMessage.walletNotReadyError), false));
+    return;
+  }
+
+  // validate address belongs to the wallet
+  const isAddressMine = yield call(wallet.isAddressMine.bind(wallet), address);
+  if (!isAddressMine) {
+    yield put(nanoContractRegisterFailure(failureMessage.addressNotMine));
+    return;
+  }
+
+  // validate nanocontract exists
+  const { ncState, error } = yield call(getNanoContractState, ncId)
+  if (error) {
+    if (error instanceof NanoRequest404Error) {
+      yield put(nanoContractRegisterFailure(failureMessage.nanoContractStateNotFound));
+    } else {
+      yield put(nanoContractRegisterFailure(failureMessage.nanoContractStateFailure));
+    }
+    return;
+  }
+
+  let ncToRegister = {
+    ncId,
+    blueprintId: ncState.blueprint_id,
+    blueprintName: ncState.blueprint_name,
+    addresses: new Set(),
+  };
+  const registeredNc = yield call(storage.getNanoContract, ncId);
+  if (registeredNc) {
+    ncToRegister = cloneDeep(registeredNc);
+  }
```

Comment:
If the nano is already registered in the wallet, an error should be thrown. The nano can't be registered with more than one address.

Reply:
Same as the above. Many addresses can register the same Nano Contract. Therefore, a Nano Contract can be already registered, but an address can't reference it twice. That's why we use the Set on `addresses` property.